Query Refactoring: Make parsing nested queries always return query builder

cbuescher opened the pull request:
Currently parsing inner queries can return `null` which leads to unnecessary complicated
null checks further up the query hierarchy. By introducing a special EmptyQueryBuilder
that can be used to signal that this query clause should be ignored upstream where possible,
we can avoid additional null checks in parent query builders and still allow for this query
to be ignored when creating the lucene query later.

This new builder returns `null` when calling its `toQuery()` method, so at this point
we still need to handle it, but having an explicit object for the intermediate query
representation makes it easier to differentiate between cases where inner query clauses are
empty (legal) or are not set at all (illegal).

This PR goes agains the query-refactoring branch.

MaineC commented:
+1 from my side to the idea and the amount of handling special cases that this PR removes.

javanna commented on core/src/main/java/org/elasticsearch/index/query/EmptyQueryBuilder.java:
```diff
@@ -0,0 +1,83 @@
+ * The current DSL allows parsing inner queries / filters like "{ }", in order to have a
+ * valid non-null representation of these clauses that actually do nothing we can use this class.
+ *
+ * This builder has no corresponding parse because it should not be allowed in the query DSL. It can
```
s/parse/parser

javanna replied:
I think saying that it should not be allowed in the query DSL is a bit misleading, cause it is allowed and we parse it properly. I know what you mean though and why you wrote that, I need yet to come up with a better explanation for this...

javanna commented on core/src/main/java/org/elasticsearch/index/query/EmptyQueryBuilder.java:
```diff
@@ -0,0 +1,83 @@
+    public static final String NAME = "empty_query";
+
+    /** the one and only empty query builder */
+    public static final EmptyQueryBuilder PROTOTYPE = new EmptyQueryBuilder();
```
I think it could be package private?

cbuescher replied:
Tried that, but I need to register the prototype in the NamedWritable registry which is in another package. Also the other builders have public prototype field.

javanna replied:
I see, bummer, I guess we can't do much about it ;)

javanna commented on core/src/main/java/org/elasticsearch/index/query/EmptyQueryBuilder.java:
```diff
@@ -0,0 +1,83 @@
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        // empty
```
is empty the same as {} ? I never know if start object and end object should be here or handled in the caller method.

cbuescher replied:
The `toXContent`wrapper around this will already render the json object brackets. If we also render them here, we end up with `{ {} }` which will not parse.

javanna replied:
cool thanks for the explanation!

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoostingQueryBuilder.java:
```diff
@@ -146,10 +145,7 @@ public String getName() {
-        if (positiveQuery == null || negativeQuery == null) {
-            return null;
-        }
+        // if either of inner query builders returns null-Query
```
I wonder if these two lines of comment still make sense, maybe we could drop them?

cbuescher replied:
They refer to the next null-checks now that are still necessary. Although the nested query builders now cannot be null anymore, they can still emit null lucene-queries. Will try to rephrase the comment.

javanna replied:
I see, then move the comment down two lines?

javanna commented on core/src/main/java/org/elasticsearch/index/query/EmptyQueryBuilder.java:
```diff
@@ -0,0 +1,83 @@
+ * be used internally when parsing nested queries to signal parent queries that the this query should be
+ * ignored upstream.
+ */
+public class EmptyQueryBuilder extends AbstractQueryBuilder<EmptyQueryBuilder> {
```
I think empty might be misleading, maybe NoOp would be better?

cbuescher replied:
I was thinking about NoOp, but the query is not really an operation, an from something called NoOp I'd expect it to be ignored completely in its usual context. This builder returns `null`-Query in `toQuery()` which needs to be handled, thats why I think NoOp is a bit misleading. I initally called it NullQueryBuilder, but that sounds too technical and a bit scary. 
It would be great if there was an empty Lucene-Query that would be ignored by enclosing queries, then we could return that instead of null in `toQuery()`, but I couldn't finde one.

javanna replied:
IgnoreMeQueryBuilder ? :D

javanna replied:
Besides jokes I see your point on NoOp naming, let's leave empty then, it doesn't convince me 100% but I cannot come up with a better name

javanna commented on core/src/main/java/org/elasticsearch/indices/query/IndicesQueriesRegistry.java:
```diff
@@ -45,6 +46,7 @@ public IndicesQueriesRegistry(Settings settings, Set<QueryParser> injectedQueryP
             }
             namedWriteableRegistry.registerPrototype(queryParser.getBuilderPrototype());
         }
+        namedWriteableRegistry.registerPrototype(EmptyQueryBuilder.PROTOTYPE);
```
maybe add a comment here on why we do this "manually" etc.

cbuescher replied:
Will do, I wasn't sure if this was the best place to register it, but I didn't want to add it to the parser registry since we should not allow top-level queries with that name.

javanna replied:
good thinking, I think it's the right place

javanna commented:
I like this a lot, left a few comments

cbuescher commented:
@javanna addressed your comments, have a second look if you like.

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoostingQueryBuilder.java:
```diff
@@ -44,9 +44,9 @@
+    private QueryBuilder positiveQuery = EmptyQueryBuilder.PROTOTYPE;
 
-    private QueryBuilder negativeQuery;
+    private QueryBuilder negativeQuery = EmptyQueryBuilder.PROTOTYPE;
```
I find a bit confusing how we handle the boosting query: the parser validates that the json contains the "negative" and "positive" elements. That said they can be empty, which would map to the new empty query builder. For java api, default is the empty query builder and one is not required to set negative and positive, no validation happens. I think that we should either require setting those two queries in both parser and builder, or remove any validation around them. 

For instance, I would go and remove the default value here, and remove validation from the parser around whether negative and positive are part of the query. Then add a check in validate method that positive and negative need to be non null. This way the behaviour stays the same as before when it comes to json (no element means no set, which means validation error), but also java api users have to remember to set the queries. Thoughts?

javanna replied:
I see how having a null default might lead to needing null checks in doXContent etc. though, let's discuss this. I think we have the opposite problem in other queries, will comment below.

javanna commented on core/src/main/java/org/elasticsearch/index/query/ConstantScoreQueryBuilder.java:
```diff
@@ -49,7 +54,7 @@
      */
     public ConstantScoreQueryBuilder(QueryBuilder filterBuilder) {
-        this.filterBuilder = filterBuilder;
+        this.filterBuilder = Objects.requireNonNull(filterBuilder);
```
it is good to check for null values, but the default is null, which means that if remove null checks (e.g. in doXContent) we might have NPEs coming from java api at least. I am on the fence on what should be the right behaviour, here too we validate that the field is present in the json, but we don't validate that the field was set through java api. We should unify our behaviour through the different queries and do the same in both parser and builder, as much as possible.

cbuescher replied:
I see what you mean. One option to guard against the NPEs you mention below would be to always initialize inner query builders to the EmptyQueryBuilder, but that way we wouldn't know if the user set it correctly or not. What if in addition to requiring non-null in constructor we also check this in validate(), then we would have to 1.) call validate() at the beginning of doXContenten() to always render legal queries 2.) validate() will be implicitely called before/after transport, so all queries should be validated once toQuery() is called. Am I missing something?

javanna replied:
I am still on the fence. I don't like null values, but I think there is a little difference between the empty query and something that was not set. That is why I don't think the default should be empty (although that might seem very convenient), which is something that requires users to put something (although incomplete) in the json. But with null values null checks are required, and relying on validate seems to be ok but not 100% correct because the java api exposes a lot of public methods that potentially throw NPE if they had no null checks, when executed against a non validated query. This last problem is not huge but it bugs me. In the end I think your last proposal around making mandatory queries required and have null checks in constructor is our best bet. We really want to remove null checks and make sure that those queries are always non null and that is the only way. But it's a breaking change for the java api, which reminds me that we might have already broken bw comp with similar changes without realising it :)

javanna commented on core/src/main/java/org/elasticsearch/index/query/FQueryFilterBuilder.java:
```diff
@@ -79,7 +83,8 @@ public String queryName() {
         builder.startObject(FQueryFilterBuilder.NAME);
-        doXContentInnerBuilder(builder, "query", queryBuilder, params);
+        builder.field("query");
+        queryBuilder.toXContent(builder, params);
```
same problem as in CostantScoreQueryBuilder I think

javanna commented on core/src/main/java/org/elasticsearch/index/query/FQueryFilterBuilder.java:
```diff
@@ -89,9 +94,6 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
-        if (this.queryBuilder == null) {
-            return null;
-        }
         Query innerQuery = this.queryBuilder.toQuery(parseContext);
```
this could lead to NPE if from the java api no set call is performed

cbuescher replied:
Actually this is not set through setter but through constructor, so it should always be okay.

javanna replied:
right I missed that good catch

javanna commented on core/src/main/java/org/elasticsearch/index/query/FieldMaskingSpanQueryBuilder.java:
```diff
@@ -99,7 +105,8 @@ public String queryName() {
         builder.startObject(NAME);
-        doXContentInnerBuilder(builder, "query", queryBuilder, params);
+        builder.field("query");
+        queryBuilder.toXContent(builder, params);
```
same as above, potential NPE?

cbuescher replied:
Set through constructor I think... true for a few others of the NPEs you mentioned, but I see the point.

javanna replied:
gotcha

javanna commented on core/src/main/java/org/elasticsearch/index/query/NotQueryBuilder.java:
```diff
@@ -79,10 +84,6 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
-        if (filter == null) {
-            return null;
-        }
-
```
potential NPEs here too I think

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryFilterBuilder.java:
```diff
@@ -60,15 +64,13 @@ public QueryBuilder innerQuery() {
         // inner query builder can potentially be `null`, in that case we ignore it
-        if (this.queryBuilder == null) {
-            return null;
-        }
```
NPEs?

cbuescher commented:
I agree that inner queries that are currently mandatory in the query DSL and are checked for in the parser should also be enforced to be set in the builder. We can do so by making all those cases mandatory arguments in the constructors now that we introduce the EmptyQueryBuilder as a stand-in for empty inner queries. I think from the cases you mentioned, only BoostingQueryBuilder needs an additional 2-arg constructor with null-checks then.

cbuescher commented:
@javanna Rebased and re-checked all the builders touched so far for breaking changes regarding inner queries. Other than the additional null-checks in constructors we're pretty consistend with the current behavior on master. Changed BoosingQueryBuilder, removing setters for inner queries and instead forcing them to be non-null in constructor, also adding this to the changes doc. Also extended handling of empty collection of inner clauses in Bool/And/OrQueryBuilder.

javanna commented on core/src/main/java/org/elasticsearch/index/query/AndQueryBuilder.java:
```diff
@@ -110,13 +110,19 @@ public Query toQuery(QueryParseContext parseContext) throws QueryParsingExceptio
+                addedFilters++;
             }
         }
+        if (addedFilters == 0) {
```
if (query.clauses().isEmpty()) ?

cbuescher replied:
Good point, didn't know this exists.

javanna commented on core/src/main/java/org/elasticsearch/index/query/AndQueryBuilder.java:
```diff
@@ -110,13 +110,19 @@ public Query toQuery(QueryParseContext parseContext) throws QueryParsingExceptio
         }
+        if (addedFilters == 0) {
+            // no inner lucene query exists, ignore upstream
+            return null;
```
not sure I get the reason behind this change. Can't we return the boolean query without clauses? What were we doing before?

cbuescher replied:
The parser so far returned null when there where no inner clauses (lucene queries). We need to do the same if none of our new builder returns a query.

javanna replied:
ah ok I think I get it, we wouldn't get here before cause what is now the empty_query was null and you could never call toQuery against it. Now we have the empty query that returns null and we have to move some null checks to the different toQuery methods.

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -294,23 +294,25 @@ public String getName() {
+        if (hasClauses()) {
+
+            BooleanQuery booleanQuery = new BooleanQuery(disableCoord);
+            int queriesAdded = addBooleanClauses(parseContext, booleanQuery, mustClauses, BooleanClause.Occur.MUST);
```
can't we replace queriesAdded with booleanQuery.clauses().size() ?

cbuescher replied:
will do.

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -294,23 +294,25 @@ public String getName() {
-        if (queryName != null) {
-            parseContext.addNamedQuery(queryName, query);
+        Query returnedQuery = new MatchAllDocsQuery();
+        if (hasClauses()) {
```
if it doesn't have clasues we don't set the boost and ignore the queryName. I think it would be a bit more readable if we had two if branches, and the set boost and named query handling after the if which kicks in in both cases.

cbuescher replied:
This will change after rebasing to cab3a68.

javanna replied:
of course how could I forget about this, I pushed it 30 mins ago :)

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -323,14 +325,17 @@ public QueryValidationException validate() {
+                queriesAdded++;
             }
         }
+        return queriesAdded;
```
if we do what I suggested above the return value shouldn't be needed.

javanna commented on core/src/main/java/org/elasticsearch/index/query/FieldMaskingSpanQueryBuilder.java:
```diff
@@ -56,7 +57,7 @@ private FieldMaskingSpanQueryBuilder() {
     public FieldMaskingSpanQueryBuilder(SpanQueryBuilder queryBuilder, String fieldName) {
         this.queryBuilder = Objects.requireNonNull(queryBuilder);
-        this.fieldName = fieldName;
+        this.fieldName = Objects.requireNonNull(fieldName);
```
why this null check for fieldName here?

cbuescher replied:
The parser checks for set fieldname, throws QueryParsingException otherwise, so we can already require it here I think.

javanna replied:
ok

javanna commented on core/src/main/java/org/elasticsearch/index/query/FieldMaskingSpanQueryParser.java:
```diff
@@ -76,9 +76,6 @@ public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOExcept
                 }
             }
         }
-        if (inner == null) {
```
inner can still be null here I'm afraid, if we never go in the if branch that parses it, that we had the queryFound flag for in other queries. I think this check is still valid

cbuescher replied:
Agreed.

javanna commented on core/src/main/java/org/elasticsearch/index/query/OrQueryBuilder.java:
```diff
@@ -107,13 +107,19 @@ public Query toQuery(QueryParseContext parseContext) throws QueryParsingExceptio
+        if (addedFilters == 0) {
+            // no inner lucene query exists, ignore upstream
+            return null;
+        }
```
here too, not sure why we have to return null now, also we can use query.clauses() to count the added filters

cbuescher replied:
Same reason as in AndQueryBuilder.

javanna commented on core/src/test/java/org/elasticsearch/index/query/QueryFilterBuilderTest.java:
```diff
@@ -50,7 +54,7 @@ protected QueryFilterBuilder createTestQueryBuilder() {
      */
     @Test
     public void testNoInnerQuery() throws QueryParsingException, IOException {
```
rename to testEmptyInnerQuery?

cbuescher replied:
okay

javanna commented on core/src/test/java/org/elasticsearch/index/query/QueryFilterBuilderTest.java:
```diff
@@ -50,7 +54,7 @@ protected QueryFilterBuilder createTestQueryBuilder() {
+        QueryFilterBuilder queryFilterQuery = new QueryFilterBuilder(EmptyQueryBuilder.PROTOTYPE);
         assertNull(queryFilterQuery.toQuery(createContext()));
     }
 
```
Maybe we should also add small tests for all of the queries where we added null checks in constructors?

javanna commented on core/src/test/java/org/elasticsearch/index/query/FieldMaskingSpanQueryBuilderTest.java:
```diff
@@ -94,5 +94,19 @@ public void testValidate() {
+        try {
+            queryBuilder = new FieldMaskingSpanQueryBuilder(new SpanTermQueryBuilder("name", "value"), null);
+            fail("Exception expected");
+        } catch (NullPointerException e) {
```
maybe move out to separate test methods and use junit expected exception rather than an empty catch?

javanna commented on core/src/test/java/org/elasticsearch/index/query/FieldMaskingSpanQueryBuilderTest.java:
```diff
@@ -83,7 +83,7 @@ public void testValidate() {
             fieldName = "fieldName";
         } else {
-            fieldName = randomBoolean() ? "" : null;
+            fieldName = "";
```
what changed here?

cbuescher replied:
We want to test validate() on a query that we can successfully construct. We have null-checks for fieldname here now.

javanna replied:
ah that's because you added a null check in the constructor for fieldName?

cbuescher replied:
yes.

javanna commented on core/src/test/java/org/elasticsearch/index/query/ConstantScoreQueryBuilderTest.java:
```diff
@@ -64,7 +70,7 @@ public void testNoFilterElement() throws IOException {
     /**
      * Test empty "filter" element.
-     * Current DSL allows inner filter element to be empty, returning a `null` inner filter builder
+     * Current DSL allows inner filter element to be empty, returning an inner NullQuerBuilder
```
something seems to be wrong with this comment? What's NullQuerBuilder?

javanna commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -223,9 +223,13 @@ public void testToQuery() throws IOException {
-        assertThat(actualQuery, instanceOf(expectedQuery.getClass()));
-        assertThat(actualQuery, equalTo(expectedQuery));
-        assertLuceneQuery(testQuery, actualQuery, context);
+        if (expectedQuery == null) {
```
can we add a small comment just to remind why we might have null here? I forget all the time :)

javanna commented:
I did another round, left a few comments

cbuescher commented:
Rebased onto the boost/queryName refactoring, had to adapt a few tests for that and also added junit tests that check constructors and setters throw exceptions where they check for `null`.

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -265,6 +265,10 @@ protected Query doToQuery(QueryParseContext parseContext) throws IOException {
 
+        if (booleanQuery.clauses().isEmpty()) {
+            return new MatchAllDocsQuery();
+        }
```
a few lines above we also have:

```
if (!hasClauses()) {
    return new MatchAllDocsQuery();
}
```

that if is against our builder clauses, for the java api I guess, while this one is for resulting lucene clauses, in case all of the parsed clasues where empty_query? Wonder if we can have a single if like that, maybe the last one alone would do it?

cbuescher replied:
I wanted to return as soon as possible but seeing that otherwise we only loop over empty collection, that should not waste much cycles. The creation of the BoolQuery would be unnecessary, but I guess if it makes the code easier to read I can change that.

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryParser.java:
```diff
@@ -71,14 +71,12 @@ public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOExcept
-                        shouldClauses.add(query);
+                    shouldClauses.add(query);
+                    if (query != EmptyQueryBuilder.PROTOTYPE) {
                         if (parseContext.isFilter() && minimumShouldMatch == null) {
```
I think i am missing something here, but when can isFilter be true here? I also wonder if the conditional here is really required or we could drop it?

cbuescher replied:
If this BoolQuery is nested inside another queries "filter" section (e.g. another BoolQuerys "filter" part), then the parent query will call `parseInnerQuery()` which will in turn switch the `isFilter` flag. So I think we need it here as soon as we add a non-null should clause.

javanna replied:
alright missed that, but do we need the `if (query != EmptyQueryBuilder.PROTOTYPE)` ?

cbuescher replied:
The original code checks the lucene query for `null` here and only sets minimumShouldMatch=1 (if not yet specified) for the first should clause added. Since we know that EmtpyQueryBuilder will not add a Should-Clause later we also don't have to set `minimumShouldMatch` yet.

javanna replied:
makes sense, add a comment that explains just that? this stuff is tricky and we will forget about it

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryParser.java:
```diff
@@ -105,14 +99,12 @@ public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOExcept
-                        if (query != null) {
-                            shouldClauses.add(query);
+                        shouldClauses.add(query);
+                        if (query != EmptyQueryBuilder.PROTOTYPE) {
```
same as above

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoostingQueryBuilder.java:
```diff
@@ -128,13 +130,10 @@ public String getName() {
         Query positive = positiveQuery.toQuery(parseContext);
         Query negative = negativeQuery.toQuery(parseContext);
+        // make upstream queries ignore this query by returning `null`
+        // if either inner query builder returns null-Query
```
s/null-query/empty query ?

cbuescher replied:
I'd prefer talking about `null` query here because we don't check for the new EmptyQueryBuilder here but any QueryBuilder returning `null` lucene query from `toQuery()`.

javanna replied:
ah I get the difference now. then maybe say "if either inner query buider returns null" null-query makes me think of the empty-query, maybe it's me

javanna commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -234,14 +234,22 @@ public void testToQuery() throws IOException {
-        assertThat(actualQuery, equalTo(expectedQuery));
-        assertLuceneQuery(testQuery, actualQuery, context);
+        // expectedQuery can be null, e.g. in case of BoostingQueryBuilder
+        // with inner clause that returns null query itself
```
s/null query/empty query ?

cbuescher replied:
same reason as above

javanna commented on core/src/test/java/org/elasticsearch/index/query/BoolQueryBuilderTest.java:
```diff
@@ -72,12 +72,14 @@ protected Query doCreateExpectedQuery(BoolQueryBuilder queryBuilder, QueryParseC
         }
 
         BooleanQuery boolQuery = new BooleanQuery(queryBuilder.disableCoord());
-        boolQuery.setBoost(queryBuilder.boost());
```
good catch!

javanna commented:
looks good, left a few comments, actually more questions than comments.

cbuescher commented:
Removed early return when creating BooleanQuery in BoolQueryBuilder and updated a few comments.

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryParser.java:
```diff
@@ -105,14 +100,13 @@ public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOExcept
-                            shouldClauses.add(query);
+                        shouldClauses.add(query);
+                        // EmptyQueryBuilder does not add lucene query later, skip setting minuminShouldMatch
+                        if (query != EmptyQueryBuilder.PROTOTYPE) {
```
other question, sorry! but I still find this confusing... previously we had null here for two situations:
1) empty filter/query
2) queries that after getting parsed would become a null query (their parse method returns null)

the first case falls now under the EmptyQueryBuilder case, which we handle here. But there can still be queries whose toQuery returns null, which need to be ignored, but that we know only in the toQuery, but we still set the minimumShouldMatch only based on case 1). I think setting the minimumShouldMatch like should happen later in the toQuery, which kinda makes sense, it needs to be done only on the resulting lucene query, not against the query builder.

So we can potentially add empty query here to the clauses as long as we move this check for null should clauses only in the toQuery method, if that makes sense

cbuescher replied:
We are still adding the EmptyQueryBuilder here in any case (before the `if`). The issue here is to set the minimumShouldMatch to at least `1`, if there is no other value supplied in the JSON _and_ we are parsing this inside another filter. This will have to happen only once for the first "real" (meaning non-null) inner query, so we can safely skip EmptyQuery here. The only thing that might go wrong here is that we set `minimumShouldMatch` to eagerly  for query != EmptyQueryBuilder but that for some reason will later be resolved to `null` in `toQuery()`. I'm not too sure about how `minimumShouldMatch` works in this case, we will have to check.

Another this that struck me just now when thinking about it: Not sure if we also need to handle this case in the  `toQuery()` itself, because query doesn't need to be parsed always in the future. The problem here is that at least for now we are using the `isFilter()` state in the QueryParseContex when using it in `toQuery()`, so there currently we don't know if the current query is used as a filter or not in the parent query. Both those issues don't have anything to do with introducing EmptyQueryBuilder here but with the fact that the decison to set `minimumShouldMatch` if not set explicitely was based on the presence of an actual lucene query so far, which really we can only know in the query-generation phase (aka toQuery). 
Since the whole conditional was moved over from the former BoolQueryFilter, it might be a good thing to ask if there is another way to solve this and why doing this only in filter context is necessary. If we could do it also for non-filter context it would be easy to move this to `toQuery`

javanna replied:
ok lets get this PR in and address this problem separately, I think we overlooked this when migrating the bool query. This check should be moved to the toQuery method. We should start looking into dividing the context into a query parse context and a to query context I think

cbuescher replied:
So should I leave the check in place for now? Can't really move it to `toQuery` as long as we have no filter context there. If I leave it, where would you suggest we track this? I'd prefer a separate issue because it's somewhat unrelated to this but complicated enough to leave some of the writup I did above there so we don't forget. Thoughts?

javanna replied:
yea I meant leave as-is and open new issue :)

javanna commented:
thanks @cbuescher left another comment :)

javanna commented:
LGTM we have to fix the bool query and probably move some logic around isFilter to the toQuery method, but let's do it in a separate issue.

cbuescher commented:
Thanks, will merge after rebase & squash and open separate issue for the BoolQueryBuilder problem mentioned above.
